data_structures/linked_list/circular_doubly_linked_list.c

```diff
@@ -0,0 +1,322 @@
+ * @brief  Function for deletion of the first node in list
+ *
+ * @param head              start pointer of list
+ * @return ListNode*        pointer to the list node after deleting first node
+ */
+ListNode *delete_from_head(ListNode *head)
+{
+    if (head == NULL)
+    {
+        printf("The list is empty\n");
+        return head;
+    }
+    ListNode *temp1, *temp2;
+    temp1 = head;
+    temp2 = temp1->prev;
+    if (temp1 == temp2)
+    {
+        free(temp2);
+        head = NULL;
+        return head;
+    }
+    temp2->next = temp1->next;
+    (temp1->next)->prev = temp2;
+    head = temp1->next;
+    free(temp1);
+    return head;
+}
+
+/**
+ * @brief Function for deletion of the last node in list
+ *
+ * @param head              start pointer of list
+ * @return ListNode*        pointer to the list node after deleting first node
+ */
+ListNode *delete_from_tail(ListNode *head)
+{
+    if (head == NULL)
+    {
+        printf("The list is empty\n");
+        return head;
+    }
+
+    ListNode *temp1, *temp2;
+    temp1 = head;
+    temp2 = temp1->prev;
+    if (temp1 == temp2)
+    {
+        free(temp2);
+        head = NULL;
+        return head;
+    }
+    (temp2->prev)->next = temp1;
+    temp1->prev = temp2->prev;
+    free(temp2);
+    return head;
+}
+
+/**
+ * @brief The function to reverse the linked list
+ *
+ * @param head              start pointer of list
+ * @return ListNode*        pointer to the list node after reversing linked list
+ */
+ListNode *reverse_list(ListNode *head)
+{
+    if (!head)
+    {
+        printf("The list is empty\n");
+        return NULL;
+    }
+    ListNode *new_head = NULL;
+    ListNode *last = head->prev;
+    ListNode *curr = last, *prev;
+
+    // traverse list in backward direction
+    while (curr->prev != last)
+    {
+        prev = curr->prev;
+        new_head = insert_at_tail(new_head, curr->value);
+        curr = prev;
+    }
+    new_head = insert_at_tail(new_head, curr->value);
+    return new_head;
+}
+
+/**
+ * @brief The function that will return current size of list
+ *
+ * @param head      start pointer of list
+ * @return int      size of list
+ */
+int getsize(ListNode *head)
+{
+    if (!head)
+    {
+        return 0;
+    }
+    int size = 1;
+    ListNode *temp = head->next;
+    while (temp != head)
+    {
+        temp = temp->next;
+        size++;
+    }
+    return size;
+}
+
+/**
+ * @brief   Display list function
+ * @param   head   start pointer of list
+ * @returns void
+ */
+
+void display_list(ListNode *head)
+{
+    printf("\nContents of your linked list: ");
+    ListNode *temp;
+    temp = head;
+    if (head != NULL)
+    {
+        while (temp->next != head)
+        {
+            printf("%lld -> ", temp->value);
+            temp = temp->next;
+        }
+        if (temp->next == head)
+        {
+            printf("%lld", temp->value);
+        }
+    }
+    else
+    {
+        printf("The list is empty");
+    }
+    printf("\n");
+}
+
```

Comment:
Here, I made an quick example for you with some of my own testing code, note that since this is just an example, some tests are not included, such as a test for `reverse_list()`:
```suggestion

/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
    ListNode *testList = NULL;
    unsigned int array[] = {2,3,4,5,6};

    assert(getsize(testList) == 0);
    
    printf("Testing inserting elements:\n");
    for (int i = 0; i < 5; ++i) {
        display_list(testList);
        testList = insert_at_head(testList,array[i]);
        assert(testList->value == array[i]);
        assert(getsize(testList) == i + 1);
    }
    
    printf("\nTesting removing elements:\n");
    for (int i = 4; i > -1; --i) {
        display_list(testList);
        assert(testList->value == array[i]);
        testList = delete_from_head(testList);
        assert(getsize(testList) == i);
    }
}

```